pkg/controller/syncset/syncset_controller_test.go

```diff
@@ -0,0 +1,313 @@
+				instanceSyncSet := &syncsetv1alpha1.SyncSet{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "syncset1",
+					},
+					Spec: syncsetv1alpha1.SyncSetSpec{
+						GVKs: []syncsetv1alpha1.GVKEntry{
+							syncsetv1alpha1.GVKEntry(podGVK),
+						},
+					},
+				}
+
+				require.NoError(t, c.Create(ctx, instanceConfig))
+				require.NoError(t, c.Create(ctx, instanceSyncSet))
+			},
+			cleanup: func(t *testing.T) {
+				t.Helper()
+
+				// reset the sync instances
+				require.NoError(t, c.Delete(ctx, instanceConfig))
+				require.NoError(t, c.Delete(ctx, instanceSyncSet1))
+			},
+			expectedGVKs: []schema.GroupVersionKind{configMapGVK, podGVK, nsGVK},
+		},
+		{
+			name: "config and multiple sync",
+			setup: func(t *testing.T) {
+				t.Helper()
+
+				instanceConfig := testutils.ConfigFor([]schema.GroupVersionKind{configMapGVK})
+				instanceSyncSet1 = &syncsetv1alpha1.SyncSet{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "syncset1",
+					},
+					Spec: syncsetv1alpha1.SyncSetSpec{
+						GVKs: []syncsetv1alpha1.GVKEntry{
+							syncsetv1alpha1.GVKEntry(podGVK),
+						},
+					},
+				}
+				instanceSyncSet2 = &syncsetv1alpha1.SyncSet{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "syncset2",
+					},
+					Spec: syncsetv1alpha1.SyncSetSpec{
+						GVKs: []syncsetv1alpha1.GVKEntry{
+							syncsetv1alpha1.GVKEntry(configMapGVK),
+						},
+					},
+				}
+
+				require.NoError(t, c.Create(ctx, instanceConfig))
+				require.NoError(t, c.Create(ctx, instanceSyncSet1))
+				require.NoError(t, c.Create(ctx, instanceSyncSet2))
+			},
+			cleanup: func(t *testing.T) {
+				t.Helper()
+
+				// reset the sync instances
+				require.NoError(t, c.Delete(ctx, instanceConfig))
+				require.NoError(t, c.Delete(ctx, instanceSyncSet1))
+				require.NoError(t, c.Delete(ctx, instanceSyncSet2))
+			},
+			expectedGVKs: []schema.GroupVersionKind{configMapGVK, podGVK},
+		},
+	}
+
+	for _, tt := range tts {
+		t.Run(tt.name, func(t *testing.T) {
+			if tt.setup != nil {
+				tt.setup(t)
+			}
+
+			assert.Eventually(t, expectedCheck(cfClient, tt.expectedGVKs), timeout, tick)
+
+			if tt.cleanup != nil {
+				tt.cleanup(t)
+
+				require.Eventually(t, func() bool {
+					return cfClient.Len() == 0
+				}, timeout, tick, "could not cleanup")
+			}
+		})
+	}
+}
+
+func expectedCheck(cfClient *fakes.FakeCfClient, expected []schema.GroupVersionKind) func() bool {
+	return func() bool {
+		for _, gvk := range expected {
+			if !cfClient.HasGVK(gvk) {
+				return false
+			}
+		}
+		return true
+	}
+}
+
+type testResources struct {
+	mgr      *manager.Manager
+	requests *sync.Map
+	cacheMgr *cm.CacheManager
+	c        *testclient.RetryClient
+	wm       *watch.Manager
+	cfClient *fakes.FakeCfClient
+	events   chan event.GenericEvent
+	cs       *watch.ControllerSwitch
+	tracker  *readiness.Tracker
+}
+
+func setupTest(ctx context.Context, t *testing.T, wrapReconciler bool, useFakeClient bool) testResources {
```

Comment:
nit: Something that would make the `setupTest` interface clearer would be to create an "options struct" that is passed in instead of arguments.

Right now, I see the following:

```
tr := setupTest(ctx, t, false, true)
```

It's unclear what false and true are.

Imagine instead:

```
ops := testOptions {
  ctx: ctx,
  testing: t,
  wrapReconciler: false,
  fakeClient: true
}
resources := setupTest(ops)
```